File: mobile_app/lib/services/task_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_constants.dart';

// Task model
class Task {
  final int taskId;
  final int careActivityId;
  final int scheduleId;
  final String gameName;
  final int gameId;
  final String time;
  final String status;

  Task({
    required this.taskId,
    required this.careActivityId,
    required this.scheduleId,
    required this.gameName,
    required this.gameId,
    required this.time,
    required this.status,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      taskId: json['taskId'] ?? 0,
      careActivityId: json['careActivityId'] ?? 0,
      scheduleId: json['scheduleId'] ?? 0,
      gameName: json['gameName'] ?? '',
      gameId: json['gameId'] ?? 0,
      time: json['time'] ?? '',
      status: json['status'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'taskId': taskId,
      'careActivityId': careActivityId,
      'scheduleId': scheduleId,
      'gameName': gameName,
      'gameId': gameId,
      'time': time,
      'status': status,
    };
  }
}

// API Result wrapper
class ApiResult<T> {
  final bool success;
  final T? data;
  final String message;
  final int? statusCode;

  ApiResult({
    required this.success,
    this.data,
    required this.message,
    this.statusCode,
  });
}

// Task Service for API calls
class TaskService {
  /// Get all games for dropdown
  static Future<ApiResult<List<Map<String, dynamic>>>> getGames() async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConstants.baseUrl}/api/games'),
        headers: {'Content-Type': 'application/json'},
      );
      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        final List<Map<String, dynamic>> games = data
            .map<Map<String, dynamic>>(
              (game) => {
                'gameid': game['gameid'],
                'name': game['name'],
                'description': game['description'],
              },
            )
            .toList();
        return ApiResult<List<Map<String, dynamic>>>(
          success: true,
          data: games,
          message: 'Games retrieved successfully',
          statusCode: response.statusCode,
        );
      } else {
        final errorData = jsonDecode(response.body);
        return ApiResult<List<Map<String, dynamic>>>(
          success: false,
          message: errorData['error'] ?? 'Failed to get games',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      print('TaskService - Error getting games: $e');
      return ApiResult<List<Map<String, dynamic>>>(
        success: false,
        message: 'Network error: ${e.toString()}',
      );
    }
  }

  /// Edit game task (game name, time, description)
  static Future<ApiResult<Task>> editGameTask({
    required int taskId,
    required String gameName,
    required String time,
    String? description,
  }) async {
    try {
      final response = await http.put(
        Uri.parse('${ApiConstants.baseUrl}/api/tasks/$taskId'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'gameName': gameName,
          'time': time,
          'description': description,
        }),
      );
      if (response.statusCode == 200) {
        final Map<String, dynamic> data = jsonDecode(response.body);
        final task = Task.fromJson(data);
        return ApiResult<Task>(
          success: true,
          data: task,
          message: 'Game task updated successfully',
          statusCode: response.statusCode,
        );
      } else {
        final errorData = jsonDecode(response.body);
        return ApiResult<Task>(
          success: false,
          message: errorData['error'] ?? 'Failed to update game task',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      print('TaskService - Error editing game task: $e');
      return ApiResult<Task>(
        success: false,
        message: 'Network error: ${e.toString()}',
      );
    }
  }

  /// Create a new task
  static Future<ApiResult<Task>> createTask({
    required int scheduleId,
    required String gameName,
    required String time,
  }) async {
    try {
      print(
        'TaskService - Creating task: scheduleId=$scheduleId, gameName=$gameName, time=$time',
      );

      final response = await http.post(
        Uri.parse('${ApiConstants.baseUrl}/api/tasks'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'scheduleId': scheduleId,
          'gameName': gameName,
          'time': time,
        }),
      );

      print('TaskService - Response status: ${response.statusCode}');
      print('TaskService - Response body: ${response.body}');

      if (response.statusCode == 200) {
        final Map<String, dynamic> data = jsonDecode(response.body);
        final task = Task.fromJson(data);

        return ApiResult<Task>(
          success: true,
          data: task,
          message: 'Task created successfully',
          statusCode: response.statusCode,
        );
      } else {
        final errorData = jsonDecode(response.body);
        return ApiResult<Task>(
          success: false,
          message: errorData['error'] ?? 'Failed to create task',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      print('TaskService - Error creating task: $e');
      return ApiResult<Task>(
        success: false,
        message: 'Network error: ${e.toString()}',
      );
    }
  }

  /// Get tasks by schedule ID
  static Future<ApiResult<List<Task>>> getTasksByScheduleId(
    int scheduleId,
  ) async {
    try {
      print('TaskService - Getting tasks for schedule ID: $scheduleId');

      final response = await http
          .get(
            Uri.parse('${ApiConstants.baseUrl}/api/tasks/schedule/$scheduleId'),
            headers: {'Content-Type': 'application/json'},
          )
          .timeout(Duration(seconds: 10));

      print('TaskService - Response status: ${response.statusCode}');
      print('TaskService - Response body: ${response.body}');

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        final List<Task> tasks = data
            .map((json) => Task.fromJson(json))
            .toList();

        return ApiResult<List<Task>>(
          success: true,
          data: tasks,
          message: 'Tasks retrieved successfully',
          statusCode: response.statusCode,
        );
      } else {
        final errorData = jsonDecode(response.body);
        return ApiResult<List<Task>>(
          success: false,
          message: errorData['error'] ?? 'Failed to get tasks',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      print('TaskService - Error getting tasks: $e');
      return ApiResult<List<Task>>(
        success: false,
        message: 'Network error: ${e.toString()}',
      );
    }
  }

  /// Update task status
  static Future<ApiResult<Task>> updateTaskStatus(
    int taskId,
    String status,
  ) async {
    try {
      print('TaskService - Updating task $taskId status to: $status');

      final response = await http.put(
        Uri.parse('${ApiConstants.baseUrl}/api/tasks/$taskId/status'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'status': status}),
      );

      print('TaskService - Response status: ${response.statusCode}');
      print('TaskService - Response body: ${response.body}');

      if (response.statusCode == 200) {
        final Map<String, dynamic> data = jsonDecode(response.body);
        final task = Task.fromJson(data);

        return ApiResult<Task>(
          success: true,
          data: task,
          message: 'Task status updated successfully',
          statusCode: response.statusCode,
        );
      } else {
        final errorData = jsonDecode(response.body);
        return ApiResult<Task>(
          success: false,
          message: errorData['error'] ?? 'Failed to update task status',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      print('TaskService - Error updating task status: $e');
      return ApiResult<Task>(
        success: false,
        message: 'Network error: ${e.toString()}',
      );
    }
  }

  /// Delete a task
  static Future<ApiResult<String>> deleteTask(int taskId) async {
    try {
      print('TaskService - Deleting task: $taskId');

      final response = await http.delete(
        Uri.parse('${ApiConstants.baseUrl}/api/tasks/$taskId'),
        headers: {'Content-Type': 'application/json'},
      );

      print('TaskService - Response status: ${response.statusCode}');
      print('TaskService - Response body: ${response.body}');

      if (response.statusCode == 200) {
        return ApiResult<String>(
          success: true,
          data: 'Task deleted successfully',
          message: 'Task deleted successfully',
          statusCode: response.statusCode,
        );
      } else {
        final errorData = jsonDecode(response.body);
        return ApiResult<String>(
          success: false,
          message: errorData['error'] ?? 'Failed to delete task',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      print('TaskService - Error deleting task: $e');
      return ApiResult<String>(
        success: false,
        message: 'Network error: ${e.toString()}',
      );
    }
  }
}

File: backend/src/main/java/Memora/DimensiaCareApplication/service/TaskService.java
package Memora.DimensiaCareApplication.service;

import Memora.DimensiaCareApplication.model.CareActivity;
import Memora.DimensiaCareApplication.model.CareActivityStatus;
import Memora.DimensiaCareApplication.model.Game;
import Memora.DimensiaCareApplication.model.Schedule;
import Memora.DimensiaCareApplication.model.Task;
import Memora.DimensiaCareApplication.repository.CareActivityRepository;
import Memora.DimensiaCareApplication.repository.GameRepository;
import Memora.DimensiaCareApplication.repository.ScheduleRepository;
import Memora.DimensiaCareApplication.repository.TaskRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@Transactional
public class TaskService {
    /**
     * Edit game task (game name, time, description)
     */
    public Map<String, Object> editGameTask(Long taskId, String gameName, String timeString, String description) {
        try {
            Task task = taskRepository.findById(taskId.intValue())
                    .orElseThrow(() -> new RuntimeException("Task not found with ID: " + taskId));

            // Update game
            Game game = gameRepository.findByName(gameName)
                    .orElseThrow(() -> new RuntimeException("Game not found with name: " + gameName));
            task.setGame(game);

            // Update time in CareActivity
            CareActivity careActivity = task.getCareActivity();
            careActivity.setTime(LocalTime.parse(timeString));

            // Optionally update description if CareActivity has a description field
            // If not, you may want to add it to the Task or CareActivity model
            // Example: careActivity.setDescription(description);
            // If not present, skip this line

            careActivityRepository.save(careActivity);
            taskRepository.save(task);

            Map<String, Object> response = convertToResponse(task);
            response.put("message", "Game task updated successfully");
            return response;
        } catch (Exception e) {
            throw new RuntimeException("Failed to edit game task: " + e.getMessage(), e);
        }
    }

    @Autowired
    private TaskRepository taskRepository;

    @Autowired
    private CareActivityRepository careActivityRepository;

    @Autowired
    private GameRepository gameRepository;

    @Autowired
    private ScheduleRepository scheduleRepository;

    /**
     * Create a new task - updates both careactivity and task tables
     */
    public Map<String, Object> createTask(Integer scheduleId, String gameName, String timeString) {
        try {
            System.out.println("TaskService - Creating task: scheduleId=" + scheduleId +
                    ", gameName=" + gameName + ", time=" + timeString);

            // Find the schedule
            Schedule schedule = scheduleRepository.findById(scheduleId.longValue())
                    .orElseThrow(() -> new RuntimeException("Schedule not found with ID: " + scheduleId));

            // Find the game by name
            Game game = gameRepository.findByName(gameName)
                    .orElseThrow(() -> new RuntimeException("Game not found with name: " + gameName));

            // Parse time string to LocalTime
            LocalTime time = LocalTime.parse(timeString);

            // Create CareActivity first
            CareActivity careActivity = new CareActivity();
            careActivity.setSchedule(schedule);
            careActivity.setTime(time);
            careActivity.setStatus(CareActivityStatus.PENDING);
            careActivity.setSkipReason(null);

            // Save care activity
            CareActivity savedCareActivity = careActivityRepository.save(careActivity);
            System.out.println("TaskService - CareActivity created with ID: " + savedCareActivity.getCareActivityId());

            // Create Task
            Task task = new Task();
            task.setCareActivity(savedCareActivity);
            task.setGame(game);

            // Save task
            Task savedTask = taskRepository.save(task);
            System.out.println("TaskService - Task created with ID: " + savedTask.getTaskId());

            // Create response
            Map<String, Object> response = new HashMap<>();
            response.put("taskId", savedTask.getTaskId());
            response.put("careActivityId", savedCareActivity.getCareActivityId());
            response.put("scheduleId", scheduleId);
            response.put("gameName", game.getName());
            response.put("gameId", game.getGameId());
            response.put("time", timeString);
            response.put("status", savedCareActivity.getStatus().toString());
            response.put("message", "Task created successfully");

            System.out.println("TaskService - Task created successfully: " + response);
            return response;

        } catch (Exception e) {
            System.err.println("TaskService - Error creating task: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Failed to create task: " + e.getMessage(), e);
        }
    }

    /**
     * Get all tasks by schedule ID
     */
    public List<Map<String, Object>> getTasksByScheduleId(Integer scheduleId) {
        try {
            System.out.println("TaskService - Getting tasks for schedule ID: " + scheduleId);

            List<Task> tasks = taskRepository.findByScheduleId(scheduleId);
            System.out.println("TaskService - Found " + tasks.size() + " tasks");

            return tasks.stream()
                    .map(this::convertToResponse)
                    .collect(Collectors.toList());

        } catch (Exception e) {
            System.err.println("TaskService - Error getting tasks: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Failed to get tasks: " + e.getMessage(), e);
        }
    }

    /**
     * Update task status (updates CareActivity status)
     */
    public Map<String, Object> updateTaskStatus(Long taskId, String status) {
        try {
            System.out.println("TaskService - Updating task " + taskId + " status to: " + status);

            Task task = taskRepository.findById(taskId.intValue())
                    .orElseThrow(() -> new RuntimeException("Task not found with ID: " + taskId));

            // Update the CareActivity status
            CareActivity careActivity = task.getCareActivity();
            careActivity.setStatus(CareActivityStatus.valueOf(status.toUpperCase()));
            careActivityRepository.save(careActivity);

            Map<String, Object> response = convertToResponse(task);
            System.out.println("TaskService - Task updated successfully: " + response);
            return response;

        } catch (Exception e) {
            System.err.println("TaskService - Error updating task: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Failed to update task: " + e.getMessage(), e);
        }
    }

    /**
     * Delete a task
     */
    public void deleteTask(Long taskId) {
        try {
            System.out.println("TaskService - Deleting task: " + taskId);

            Task task = taskRepository.findById(taskId.intValue())
                    .orElseThrow(() -> new RuntimeException("Task not found with ID: " + taskId));

            taskRepository.delete(task);
            System.out.println("TaskService - Task deleted successfully");

        } catch (Exception e) {
            System.err.println("TaskService - Error deleting task: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Failed to delete task: " + e.getMessage(), e);
        }
    }

    /**
     * Convert Task entity to response map
     */
    private Map<String, Object> convertToResponse(Task task) {
        CareActivity careActivity = task.getCareActivity();
        Game game = task.getGame();

        Map<String, Object> response = new HashMap<>();
        response.put("taskId", task.getTaskId());
        response.put("careActivityId", careActivity.getCareActivityId());
        response.put("scheduleId", careActivity.getSchedule().getScheduleId());
        response.put("gameName", game.getName());
        response.put("gameId", game.getGameId());
        response.put("time", careActivity.getTime().toString());
        response.put("status", careActivity.getStatus().toString());

        return response;
    }
}
